File: backend/src/providers/menuItemProvider.ts
import { MenuItemUpdateBodyType } from '../schemas/menuItem';
import { prisma } from '../lib/prisma.js';
import { MenuItemCreateBodyType } from '../schemas/menuItem.js';
import { BaseSearchRequestType } from '../schemas/search.js';
import parseFilterString from '../utils/filterParser.js';

/**
 * Data Access Layer — MenuItem
 */
export const menuItemProvider = {
  /** Return all active menu items, optionally filtered by category name. */
  async findAll(request: BaseSearchRequestType) {
    const where = parseFilterString(request.search, {
      allowedFields: [
        'id',
        'name',
        'description',
        'price',
        'deleted',
        'active',
        'category_id',
        'rating',
        'tag',
      ],
      defaultSearchFields: ['name', 'description'],
      defaultDeleted: true,
    });

    if (request.categoryId) {
      (where as Record<string, unknown>).category_id = request.categoryId;
    }

    return prisma.menuItem.findMany({
      where,
      include: {
        category: true,
        menu_item_images: {
          where: { is_primary: true },
          take: 1,
        },
      },
      orderBy: { id: request.order },
      take: request.limit + 1,
      ...(request.cursor ? { cursor: { id: request.cursor }, skip: 1 } : {}),
    });
  },

  /** Fetch specific menu items by their IDs (used when creating orders). */
  async findByIds(ids: bigint[]) {
    return prisma.menuItem.findMany({
      where: { id: { in: ids }, deleted: false, active: true },
    });
  },

  async findById(id: number) {
    return prisma.menuItem.findFirst({
      where: { id, deleted: false, active: true },
      include: {
        category: true,
        menu_item_images: {
          orderBy: [{ is_primary: 'desc' }, { display_order: 'asc' }],
        },
      },
    });
  },

  async create(data: MenuItemCreateBodyType) {
    const payload: any = { ...data };
    // remove client-only/unknown fields that are not in Prisma model
    if (payload.image != null) delete payload.image;
    if (payload.category_id != null) {
      payload.category = { connect: { id: BigInt(payload.category_id) } };
      delete payload.category_id;
    }
    return prisma.menuItem.create({ data: payload });
  },

  async update(data: MenuItemUpdateBodyType, id: number) {
    const payload: any = { ...data };
    // remove client-only/unknown fields
    if (payload.image != null) delete payload.image;
    if (Object.prototype.hasOwnProperty.call(payload, 'category_id')) {
      const cid = payload.category_id;
      delete payload.category_id;
      if (cid == null) {
        payload.category = { disconnect: true };
      } else {
        payload.category = { connect: { id: BigInt(cid) } };
      }
    }

    return prisma.menuItem.update({ where: { id }, data: payload });
  },

  async delete(id: number) {
    return prisma.menuItem.update({
      where: { id },
      data: { deleted: true },
    });
  },
};

File: backend/app/providers/menu_item_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, func
from sqlalchemy.orm import selectinload
from typing import Optional, List
from app.models.menu_item import MenuItem
from app.models.menu_item_image import MenuItemImage
from app.models.category import Category

class MenuItemProvider:
    async def find_all(self, db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None, category_id: Optional[int] = None):
        from sqlalchemy import asc, desc
        stmt = (
            select(MenuItem)
            .options(
                selectinload(MenuItem.category),
                selectinload(MenuItem.menu_item_images),
            )
            .where(MenuItem.deleted == False)
        )
        if search:
            stmt = stmt.where(MenuItem.name.ilike(f"%{search}%") | MenuItem.description.ilike(f"%{search}%"))
        if category_id:
            stmt = stmt.where(MenuItem.category_id == category_id)
        if cursor:
            if order == "desc":
                stmt = stmt.where(MenuItem.id < cursor)
            else:
                stmt = stmt.where(MenuItem.id > cursor)
        stmt = stmt.order_by(desc(MenuItem.id) if order == "desc" else asc(MenuItem.id))
        stmt = stmt.limit(limit + 1)
        result = await db.execute(stmt)
        return result.scalars().all()

    async def find_by_ids(self, db: AsyncSession, ids: List[int]) -> List[MenuItem]:
        result = await db.execute(
            select(MenuItem).where(MenuItem.id.in_(ids), MenuItem.deleted == False, MenuItem.active == True)
        )
        return list(result.scalars().all())

    async def find_by_id(self, db: AsyncSession, item_id: int) -> Optional[MenuItem]:
        result = await db.execute(
            select(MenuItem)
            .options(selectinload(MenuItem.category), selectinload(MenuItem.menu_item_images))
            .where(MenuItem.id == item_id, MenuItem.deleted == False, MenuItem.active == True)
        )
        return result.scalar_one_or_none()

    async def create(self, db: AsyncSession, **data) -> MenuItem:
        item = MenuItem(**data)
        db.add(item)
        await db.commit()
        await db.refresh(item)
        return item

    async def update(self, db: AsyncSession, item_id: int, **data) -> Optional[MenuItem]:
        await db.execute(update(MenuItem).where(MenuItem.id == item_id).values(**data))
        await db.commit()
        return await self.find_by_id(db, item_id)

    async def soft_delete(self, db: AsyncSession, item_id: int) -> Optional[MenuItem]:
        await db.execute(update(MenuItem).where(MenuItem.id == item_id).values(deleted=True))
        await db.commit()
        result = await db.execute(select(MenuItem).where(MenuItem.id == item_id))
        return result.scalar_one_or_none()

    async def update_average_rating(self, db: AsyncSession, menu_item_id: int) -> None:
        from app.models.menu_item_comment import MenuItemComment
        result = await db.execute(
            select(func.avg(MenuItemComment.rating)).where(
                MenuItemComment.menu_item_id == menu_item_id,
                MenuItemComment.deleted == False,
                MenuItemComment.rating.isnot(None),
            )
        )
        avg_rating = result.scalar_one_or_none()
        await db.execute(
            update(MenuItem).where(MenuItem.id == menu_item_id).values(rating=avg_rating)
        )
        await db.commit()

menu_item_provider = MenuItemProvider()
